### ISSUE TITLE
Add support for history states in Umple

### ISSUE THREAD

**Issue by Nava2** [umple/umple#138]
Originally reported on Google Code with ID 238
Owned by `erictelmer`

---

Umple currently does not support history states.

Suggested keywords: hist, and history.

Example:

eventName -> hist.parentStateName;

---

Reported by @obahy on 2011-10-12 17:17:16

**Comment by Nava2** [comment]
Remarks:
- At most one history state in a composite state machine.
- If the composite state machine has never been active, then the transition to the
  history state does not take place (and the event handling function shall return false.)
  unless there is a transition from the history state to another state.

---

Reported by @obahy on 2011-10-12 17:26:26

**Comment by Nava2** [comment]
Reported by @umple on 2011-10-13 19:25:34
- **Labels added**: Component-Syntax

**Comment by Nava2** [comment]
Reported by @umple on 2012-06-14 21:59:24
- **Labels added**: Diffic-Med

**Comment by Nava2** [comment]
This example demonstrates the semantics of History state. This state machine has two
states, Play and Stop. Play state has two sub-states, On and Off. When inside the Stop
state, and resume event occurs, a transition to history state takes place. The value
of history state is the last configuration of the composite state Play. This is to
say that a transition to history state takes the state machine to the last active sub-state
(either On or Off in this example). For deep history (see issue 139).

class HistoryState {
  status {
    Play {
      On { 
        turnOff -> Off;
        stop -> Stop;
      }
      Off { 
        turnOn -> On;
        stop -> Stop; 
      }
    }
    Stop {
      resume -> History; // this transition takes the state machine to either On or
Off (depending on which one was the last active state)
    }
  }
}

---

Reported by @obahy on 2012-06-19 15:30:03

**Comment by Nava2** [comment]
Reported by @umple on 2012-11-22 20:34:33
- **Labels added**: Type-Enhancement, Priority-Low
- **Labels removed**: Type-Defect, Priority-Medium

**Comment by Nava2** [comment]
see also issue 139

---

Reported by @umple on 2013-02-07 17:35:58
- **Labels added**: Priority-Medium
- **Labels removed**: Priority-Low

**Comment by Nava2** [comment]
Reported by @umple on 2014-01-31 19:00:00
- **Owner removed**
- **cc added:** +@obahy

**Comment by Nava2** [comment]
Reported by @umple on 2014-02-22 14:05:44
- **Blocked on**: #417

**Comment by Nava2** [comment]
Reported by @umple on 2014-02-22 15:13:38
- **Blocking**: #139

**Comment by Nava2** [comment]
Please ignore the details earlier. We will not be using 'hist' etc.

I am adding some details of what needs to be done for history states (issue 138 and
issue 139). Both should be done together, since the difference between them is minor.

  a) There should be only minimal need for new syntax (changes to grammar) . Simply
use H (capital) as a reserved word state name for history, and H\* as a reserved word
state machine for deep history. This is consistent with UML. Since the \* can't be part
of a normal state name, this would have to be added to the grammar to allow it to be
valid (both in transitions and in declaration of states). Allowing this syntax could
be a first step in agile introduction of this feature.

  b) If there is a composite state s, then transitions should be able to transition
to s.H or s.H\* where s is any composite state. Transitioning to H (as opposed to s.H)
would not make much sense since such transitions are only useful when transitioning
to a different composite state.

  c) The destination of a H or H\* transition could be a distant composite state, or
a parent in the state hierarchy, or a sibling or nephew composite state.

  d) There can be H and H\* at multiple levels of a composite state hierarchy.

  e) H or H\* in a non-composite state would not be allowed and would result in an error.
H\* in a one-level state would be allowed and would be equivalent to H.

  f) The state H and H\* does not need to actually exist in the destination composite
state s. If it is not there, then no warning would be raised and it would be created
anyway.

  g) H and H\* can both be created in any composite state. They would only be valid
if somewhere there is a transition to them. Otherwise there would be a warning. Issuing
this warning is lower priority.

  h) If H or H\* does exist then it would not be allowed to have any event transitions,
nor any do activity. Presence of these would result in an error. Only an auto-transition
would be allowed to specify the situation where there is a transition to it but the
composite state has never been visited before.

  i) To record history, there needs to be a 'lastVisitedState' instance variable in
a state machine that is set on entry into any state to record the history. This technically
only needs to be set if there is the possibility of transition to history, but it might
be simpler initially to always set it.

   j) Semantics for processing transitions 
    - On taking a transition to H: 
       i. Deal with exit actions of the state being left as normal.
       ii. Go to the H state.
       iii. Process any entry then any exit actions
       iv. If there is history (lastVisitedState is visited) go to that substate in
the normal way, as if taking an auto-transition.
       v. If there is no history, but there is an auto-transition specified in the
H state, then take it.
       vi Otherwise go to the default start state of the composite state and follow
normal processing.
       vii. Deal with entry into the new destination as normal

```
- On taking a transition to H*
   i. Deal with exit actions of the state being left as normal.
   ii. Go to the H* state.
   iii. Process any entry then any exit actions
   iv. If there is history (lastVisitedState is visited) iterate through substates
```

until the leaf is reached to find the leaf state last visited and go to that leaf state
in the normal way, as if taking an auto-transition.
       v. If there is no history, but there is an auto-transition specified in the
H state, then take it.
       vi Otherwise go to the default start state of the composite state and follow
normal processing.
       vii. Deal with entry into the new destination as normal

Below is an example that can be one of the many test cases for this. We need syntax
and testbed tests

class X {
  sm {
    safe {
      alert -> emergency;
      H {}
      normal {
        s1 {
        }
        s2 {
        }
        s3 {
        }
      }
      closed {
        regularClosed {
        }
        maintenance {
        }
      }
    }
    emergency {
      allClear -> safe.H;
    }
  }
}

---

Reported by @umple on 2014-02-22 15:42:15
- **Owner changed**: `erictelmer`

**Comment by Nava2** [comment]
Reported by @umple on 2014-05-12 13:00:57
- **Status changed**: `Verified`
